test/js/applesoft/compiler.spec.ts

```diff
@@ -0,0 +1,431 @@
+// assembler, revision 4/27/84. There is evidence from
+// https://www.pagetable.com/?p=774 that the original Microsoft
+// BASIC source code used these names as well.
+const TXTTAB = 0x67; // start of program, word
+const VARTAB = 0x69; // start of variables, word
+const ARYTAB = 0x6B; // start of arrays, word
+const STREND = 0x6D; // end of strings, word
+const PRGEND = 0xAF; // end of program, word
+
+function compileToMemory(ram: Memory, program: string) {
+    ApplesoftCompiler.compileToMemory(ram, program);
+}
+
+// Manual decompilation based on "Applesoft Internal Structure"
+// by C.K. Mesztenyi/Washington Apple Pi, from Call—A.P.P.L.E.,
+// January, 1982. Archived at:
+// https://archive.org/details/DTCA2DOC-045_applesoft_internal
+// Decompilation verified on the emulator by typing in the
+// program, then:
+//    ]CALL -151
+//    *800.820
+// and comparing the resulting bytes (starting at 801).
+
+describe('ApplesoftCompiler', () => {
+    it('compiles a one-line hello world', () => {
+        const compiler = new ApplesoftCompiler();
+        compiler.compile('10 PRINT "HELLO, WORLD!"');
+        expect(compiler.program()).toEqual(new Uint8Array([
+            0x16, 0x08, 0x0a, 0x00, 0xba, 0x22, 0x48, 0x45,
+            0x4c, 0x4c, 0x4f, 0x2c, 0x20, 0x57, 0x4f, 0x52,
+            0x4c, 0x44, 0x21, 0x22, 0x00, 0x00, 0x00
+        ]));
+    });
+
+    it('compiles a one-line hello world into memory', () => {
+        const ram = new RAM(0, 0xff); // 64K of RAM
+
+        compileToMemory(ram, '10 PRINT "HELLO, WORLD!"');
+        expect(ram.read(0x08, 0x01)).toBe(0x16); // pointer to next line low
+        expect(ram.read(0x08, 0x02)).toBe(0x08); // pointer to next line high
+        expect(ram.read(0x08, 0x03)).toBe(10);   // line number low
+        expect(ram.read(0x08, 0x04)).toBe(0);    // line number high
+        expect(ram.read(0x08, 0x05)).toBe(0xba); // PRINT
+        expect(ram.read(0x08, 0x06)).toBe(0x22); // "
+        expect(ram.read(0x08, 0x07)).toBe(0x48); // H
+        expect(ram.read(0x08, 0x08)).toBe(0x45); // E
+        expect(ram.read(0x08, 0x09)).toBe(0x4C); // L
+        expect(ram.read(0x08, 0x0a)).toBe(0x4C); // L
+        expect(ram.read(0x08, 0x0b)).toBe(0x4F); // O
+        expect(ram.read(0x08, 0x0c)).toBe(0x2C); // ,
+        expect(ram.read(0x08, 0x0d)).toBe(0x20); // space
+        expect(ram.read(0x08, 0x0e)).toBe(0x57); // W
+        expect(ram.read(0x08, 0x0f)).toBe(0x4F); // O
+        expect(ram.read(0x08, 0x10)).toBe(0x52); // R
+        expect(ram.read(0x08, 0x11)).toBe(0x4C); // L
+        expect(ram.read(0x08, 0x12)).toBe(0x44); // D
+        expect(ram.read(0x08, 0x13)).toBe(0x21); // !
+        expect(ram.read(0x08, 0x14)).toBe(0x22); // "
+        expect(ram.read(0x08, 0x15)).toBe(0x00); // end of line
+        expect(ram.read(0x08, 0x16)).toBe(0x00); // end of program low
+        expect(ram.read(0x08, 0x17)).toBe(0x00); // end of program high
```

Comment:
I wonder if using `TestMemory` and adding some array accessors would make these a less dense.

Reply:
This is a good point. I've been trying to figure out something reasonable, but the best I get is:

```typescript
expect(ram.range(0x0801, 0x0817)).toEqual([
    0x16, // pointer to next line low
    0x08, // pointer to next line high
    ...
]);
```

Another choice could be:

```typescript
expect(ram.range(0x0801, 0x0817)).toMatch({
    0x0801: 0x16, // pointer to next line low
    0x0801: 0x08, // pointer to next line high
    ...
});
```

Thoughts?

Reply:
Sorry I completely spaced on this - if you loop back to this I like the first version. Using the constants for tokens might also help readability.